ulib/FStar.WellFoundedRelation.fsti

```diff
@@ -0,0 +1,175 @@
+    `acc_to_wfr r f`
+
+    This is a `wfr_t` built from a relation `r` and a function `f:
+    well-founded r`.  Such a function demonstrates that `r` is
+    well-founded using the accessibility type `acc` described in
+    FStar.WellFounded.fst.  So, `precedes_by_wfr (acc_to_wfr r f) x1
+    x2` is true whenever there exists an object of type `r x1 x2`.
+
+    `subrelation_to_wfr r wfr`
+
+    This is a `wfr_t` built from a relation `r` that's a subrelation
+    of an existing well-founded relation `wfr`.  By "subrelation" we
+    mean that any pair related by `r` is also related by `wfr`.  So,
+    `precedes_by_wfr (subrelation_to_wfr r wfr) x1 x2` is equivalent
+    to `r x1 x2`.
+
+    `inverse_image_to_wfr r f wfr`
+
+    This is a `wfr_t` built from a relation `r: a -> a -> Type0`, a
+    function `f: a -> b`, and an existing well-founded relation `wfr`
+    on `b`.  The relation `r` must be an "inverse image" of `wfr`
+    using the mapping function `f`, meaning that
+    `forall x1 x2. r x1 x2 ==> precedes_by_wfr wfr (f x1) (f x2)`.
+    So, `precedes_by_wfr (inverse_image_to_wfr r f wfr) x1 x2` is
+    equivalent to `r x1 x2`.
+
+    `lex_nondep_wfr wfr_a wfr_b`
+
+    This is a `wfr_t` describing lexicographic precedence for
+    non-dependent tuples of some type `a * b`.  It's built from two
+    well-founded relations: `wfr_a` on type `a` and `wfr_b` on type
+    `b`.  So, `precedes_by_wfr (lex_nondep_wfr wfr_a wfr_b) xy1 xy2`
+    is equivalent to `let (x1, y1), (x2, y2) = xy1, xy2 in
+    precedes_by_wfr wfr_a x1 x2 \/ (x1 == x2 /\ precedes_by_wfr wfr_b
+    y1 y2)`.
+
+    `lex_dep_wfr wfr_a wfr_b`
+
+    This is a `wfr_t` describing lexicographic precedence for
+    dependent tuples of type `(x: a & b x)`.  It's built from a
+    well-founded relation `wfr_a` on type `a` and a function `wfr_b`
+    that maps each `x: a` to a `wfr_t` on type `(b x)`.  So,
+    `precedes_by_wfr (lex_dep_wfr wfr_a wfr_b) xy1 xy2` is equivalent
+    to `let (| x1, y1 |), (| x2, y2 |) = xy1, xy2 in precedes_by_wfr
+    wfr_a x1 x2 \/ (x1 == x2 /\ precedes_by_wfr (wfr_b x1) y1 y2)`.
+*)
+
+module FStar.WellFoundedRelation
+
+open FStar.WellFounded
+
+val wfr_t (a: Type u#a) (d: a -> Type u#d) : Type u#(max a d + 1)
+
+val decreaser_for_wfr (#a: Type u#a) (#d: a -> Type u#d) (wfr: wfr_t a d) (x: a) : d x
+
+val precedes_by_wfr (#a: Type u#a) (#d: a -> Type u#d) (wfr: wfr_t a d) (x1: a) (x2: a)
+  : (related: Type0{related ==> decreaser_for_wfr wfr x1 << decreaser_for_wfr wfr x2})
+
+val default_wfr (a: Type u#a) : (wfr: wfr_t a (fun _ -> a) {forall x1 x2. precedes_by_wfr wfr x1 x2 <==> x1 << x2})
+
+val empty_wfr (a: Type u#a) : (wfr: wfr_t a (fun _ -> a){forall x1 x2. ~(precedes_by_wfr wfr x1 x2)})
+
+val acc_to_wfr (#a: Type u#a) (r: a -> a -> Type0) (f: well_founded r{forall x1 x2 (p: r x1 x2). (f x2).access_smaller x1 p == f x1})
+  : (wfr: wfr_t a (fun (x: a) -> acc r x) {forall x1 x2. precedes_by_wfr wfr x1 x2 <==> (exists (p: r x1 x2). True)})
```

Comment:
FYI, I have a slightly better version of this combinator after generalized FStar.WellFounded to work with higher universes. 

```
let sq (#a:Type) (r:binrel u#a u#r a) (x y:a) = squash (r x y)

val acc_to_wfr_alt (#a: Type u#a)
                   (r: binrel u#a u#r a)
                   (f: well_founded (sq r))
  : wfr: wfr_t a (fun (x: a) -> acc r x) {
          forall x1 x2 (p: r x1 x2). precedes_by_wfr wfr x1 x2
    }
```

It depends on improvements in nik_2560

So, we can try to merge this one first and then revise it once I can land nik_2560.